datasafe-rest/src/main/java/de/adorsys/datasafe/rest/controller/DocumentController.java

```diff
@@ -0,0 +1,86 @@
+package de.adorsys.datasafe.rest.controller;
+
+import de.adorsys.datasafe.business.api.types.UserID;
+import de.adorsys.datasafe.business.api.types.UserIDAuth;
+import de.adorsys.datasafe.business.api.types.action.ListRequest;
+import de.adorsys.datasafe.business.api.types.action.ReadRequest;
+import de.adorsys.datasafe.business.api.types.action.RemoveRequest;
+import de.adorsys.datasafe.business.api.types.action.WriteRequest;
+import de.adorsys.datasafe.business.api.types.keystore.ReadKeyPassword;
+import de.adorsys.datasafe.business.api.types.resource.DefaultPrivateResource;
+import de.adorsys.datasafe.business.api.types.resource.PrivateResource;
+import de.adorsys.datasafe.business.impl.service.DefaultDatasafeServices;
+import lombok.SneakyThrows;
+import org.springframework.util.StreamUtils;
+import org.springframework.web.bind.annotation.*;
+
+import javax.servlet.http.HttpServletResponse;
+import java.io.ByteArrayOutputStream;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.URI;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static org.springframework.http.MediaType.APPLICATION_OCTET_STREAM_VALUE;
+
+@RestController
+@RequestMapping("/document")
+public class DocumentController {
+
+    private final DefaultDatasafeServices dataSafeService;
+
+    public DocumentController(DefaultDatasafeServices dataSafeService) {
+        this.dataSafeService = dataSafeService;
+    }
+
+    @SneakyThrows
+    @GetMapping(produces = APPLICATION_OCTET_STREAM_VALUE)
+    public void readDocument(@RequestParam String user,
+                             @RequestParam String password,
+                             @RequestParam String path,
+                             HttpServletResponse response) {
+        UserIDAuth userIDAuth = new UserIDAuth(new UserID(user), new ReadKeyPassword(password));
+        PrivateResource resource = DefaultPrivateResource.forPrivate(URI.create(path));
+        ReadRequest<UserIDAuth, PrivateResource> request = ReadRequest.forPrivate(userIDAuth, resource);
+        InputStream inputStream = dataSafeService.privateService().read(request);
+        OutputStream outputStream = response.getOutputStream();
+        StreamUtils.copy(inputStream, outputStream);
+        inputStream.close();
+        outputStream.close();
+    }
+
+    @SneakyThrows
+    @PostMapping(consumes = APPLICATION_OCTET_STREAM_VALUE)
+    public void writeDocument(@RequestParam String user,
+                              @RequestParam String password,
+                              @RequestParam String path,
```

Comment:
Why not to use PathVariable instead? If used as path then request can be PUT https://example.com/document/somewhere/file that is very natural
Please consider it everywhere

Reply:
had a problem with dots in path. then we need to urlencode in request

Reply:
In which request - to REST api, or to library? If it is REST api it is completely OK

Reply:
With RESt api. Especially with slashes. I come up with some workaround. Maybe make interceptor which extract path from request and pass it as requestParam?

Reply:
resolved